### ISSUE TITLE
unit test failure with migrate_gradebook.

### ISSUE THREAD

**Issue by danmarsden** [turnitin/moodle-mod_turnitintooltwo#387]
Testing suite mod_turnitintooltwo_testsuite
Moodle 3.1.8 (Build: 20170911)
Php: 7.0.22, pgsql: 9.5.9, OS: Linux 4.4.0-87-generic x86_64
PHPUnit 4.8.27 by Sebastian Bergmann and contributors.

........................................F........

Time: 28.29 seconds, Memory: 74.00MB

There was 1 failure:

1) mod_lib_testcase::test_turnitintooltwo_cron_migrate_gradebook_1000
Failed asserting that 400 matches expected 0.

/var/www/html/mod/turnitintooltwo/tests/unit/lib_test.php:141
/var/www/html/lib/phpunit/classes/advanced_testcase.php:80

To re-run:
 /var/www/html/vendor/bin/phpunit --colors mod_lib_testcase mod/turnitintooltwo/tests/unit/lib_test.php

FAILURES!
Tests: 49, Assertions: 200, Failures: 1.

**Comment by danmarsden** [comment]
I think this might occur because migrate_gradebook is not doing a sort on the request for $assignments
https://github.com/turnitin/moodle-mod_turnitintooltwo/blob/master/lib.php#L697

And the unit tests expect the assignments to be processed in the same order as they were written to the db.

I'd guess that adding a sort by "id" or "turnitintooltwoid" to the request for $assignments would probably fix it.

also - this line looks a bit weird to me:
https://github.com/turnitin/moodle-mod_turnitintooltwo/blob/master/lib.php#L703

This part of the statement doesn't look like it would ever validate:
...|| ($numsubmissions == 0) && ($assignment->numsubmissions > MIGRATION_SUBMISSIONS_CUTOFF)

because $numsubmissions is set to +$assignment->numsubmissions in the line before.

I think what might be intended is the following?
```
if ($numsubmissions <= MIGRATION_SUBMISSIONS_CUTOFF) || 
  (($numsubmissions == 0) && ($assignment->numsubmissions > MIGRATION_SUBMISSIONS_CUTOFF)) {
    $numsubmissions += $assignment->numsubmissions;
```

not the slightly different bracket configuration in the if statement - wrapping the last 2 checks in a set of brackets and moving $numsubmissions += to under the if statement instead of before.

**Comment by danmarsden** [comment]
in fact - this might be more what was intended:
```
        if (($numsubmissions + $assignment->numsubmissions) <= MIGRATION_SUBMISSIONS_CUTOFF ||
           (($numsubmissions == 0) && ($assignment->numsubmissions > MIGRATION_SUBMISSIONS_CUTOFF))) {

            $numsubmissions += $assignment->numsubmissions;
```

**Comment by dwinn** [comment]
Hi @danmarsden,

Thanks for the contribution and pull request. I have taken a look at it.

I can't get the unit test to fail myself when I run the suite but I see no harm in implementing the sort field you request.

For the second point, you've definitely found an error here with the second part of the logic check. At first glance it can be difficult to understand what we're trying to achieve here. It boils down to the Moodle calls for updating the gradebook being slow. On their own they seem quick but once the numbers get into the hundreds or thousands you begin to notice a slowdown so we decided to place the gradebook migrations in the cron and limit this to 1000 per cron run. This allows us to migrate an assignment quickly while performing slower processes in a controlled way in the cron.

We also made it so an assignment's gradebook wouldn't be left half migrated, so we total up the grades to be migrated and the grades migrated so far and determine whether to migrate the assignment's gradebook. If this total is above `MIGRATION_SUBMISSIONS_CUTOFF` (1000) we break out of the loop and don't migrate that assignment on this run of the cron.

The second part after the || is to handle assignments that contain more than 1000 grades since they'd always throw the total over 1000. The idea is if we have > 1000 grades for an assignment we'll migrate the gradebook for that assignment as a whole - ignoring the 1000 cut off. But we ensure this is the only assignment we migrate during that cron run, hence the 0 check.

Looking at your change, this fixes the issue. You are right that there is a bug in the logic since `$numsubmissions` will never be 0 and there are missing brackets in the second logic check. The code you provide will work as the `$numsubmissions` variable can evaluate to 0 during the second check but is still referenced correctly during the first check with the change you made there.

I'll merge your pull request and it will be added to our upcoming release.
